AK/HashTable.h

```diff
@@ -262,39 +348,55 @@ class HashTable {
         auto& bucket = lookup_for_writing(value);
         new (bucket.slot()) T(move(value));
         bucket.used = true;
+
+        if constexpr (IsOrdered) {
+            if (!m_housekeeping.head) [[unlikely]] {
+                m_housekeeping.head = &bucket;
+            } else {
+                bucket.previous = m_housekeeping.tail;
+                m_housekeeping.tail->next = &bucket;
+            }
+            m_housekeeping.tail = &bucket;
+        }
     }
 
     void rehash(size_t new_capacity)
     {
         new_capacity = max(new_capacity, static_cast<size_t>(4));
-        new_capacity = kmalloc_good_size(new_capacity * sizeof(Bucket)) / sizeof(Bucket);
+        new_capacity = kmalloc_good_size(new_capacity * sizeof(BucketType)) / sizeof(BucketType);
 
         auto* old_buckets = m_buckets;
-        auto old_capacity = m_capacity;
+        Iterator old_iter = begin();
+
+        if constexpr (IsOrdered) {
+            m_buckets = (BucketType*)kmalloc(sizeof(BucketType) * (new_capacity));
+            __builtin_memset(m_buckets, 0, sizeof(BucketType) * (new_capacity));
+
+            m_housekeeping = { nullptr, nullptr };
+        } else {
+            m_buckets = (BucketType*)kmalloc(sizeof(BucketType) * (new_capacity + 1));
+            __builtin_memset(m_buckets, 0, sizeof(BucketType) * (new_capacity + 1));
+        }
 
-        m_buckets = (Bucket*)kmalloc(sizeof(Bucket) * (new_capacity + 1));
-        __builtin_memset(m_buckets, 0, sizeof(Bucket) * (new_capacity + 1));
         m_capacity = new_capacity;
         m_deleted_count = 0;
 
-        m_buckets[m_capacity].end = true;
+        if constexpr (!IsOrdered)
+            m_buckets[m_capacity].end = true;
 
         if (!old_buckets)
             return;
 
-        for (size_t i = 0; i < old_capacity; ++i) {
-            auto& old_bucket = old_buckets[i];
-            if (old_bucket.used) {
-                insert_during_rehash(move(*old_bucket.slot()));
-                old_bucket.slot()->~T();
-            }
+        for (auto it = move(old_iter); it != end(); ++it) {
+            insert_during_rehash(move(*it));
+            it->~T();
```

Comment:
this is a bit ugly (and using the iterator of an actively-being destroyed object feels unsafe), how about just storing the old m_head and doing a good old
```c++
auto* old_bucket = ...;
while (old_bucket) {
    insert_during_rehash(move(*old_bucket->slot()));
    old_bucket->slot()->~T();
    old_bucket = old_bucket->next;
}
```

Reply:
but that would only work in the Ordered case, and for the unordered case the code would be the same it was before, which is pretty much a copy of the iterator implementation

On another hand  the buckets don't get destroyed until after this loop, so safety wise, this should be fine

Reply:
i guess its fine then :shrug: